model/flow/sealing_segment.go

```diff
@@ -0,0 +1,102 @@
+package flow
+
+import (
+	"fmt"
+)
+
+// SealingSegment is the chain segment such that the head (greatest
+// height) is this snapshot's reference block and the tail (least height)
+// is the most recently sealed block as of this snapshot (ie. the block
+// referenced by LatestSeal). The segment is in ascending height order.
+type SealingSegment struct {
+	// Blocks the chain segment blocks
+	Blocks []*Block
+
+	// Due to decoupling of execution receipts it's possible that blocks from sealing segment will be referring
+	// execution results incorporated in blocks that aren't part of the segment.
+	// ExecutionResults will contain those results.
+	ExecutionResults ExecutionResultList
+}
+
+func (segment *SealingSegment) Highest() *Block {
+	return segment.Blocks[len(segment.Blocks)-1]
+}
+
+func (segment *SealingSegment) Lowest() *Block {
+	return segment.Blocks[0]
+}
+
+var (
+	ErrSegmentMissingSeal = fmt.Errorf("sealing segment failed sanity check: highest block in segment does not contain seal for lowest")
+	ErrSegmentBlocksEmpty = fmt.Errorf("invalid sealing segment with 0 blocks")
+)
+
+type SealingSegmentBuilder struct {
+	resultLookup    func(resultID Identifier) (*ExecutionResult, error)
+	includedResults map[Identifier]*ExecutionResult
+	blocks          []*Block
+	results         []*ExecutionResult
+}
+
+// AddBlock appends block to blocks
+func (builder *SealingSegmentBuilder) AddBlock(block *Block) error {
+	for _, receipt := range block.Payload.Receipts {
+		if _, ok := builder.includedResults[receipt.ResultID]; ok {
+			continue
+		}
+
+		resultsByID := block.Payload.Results.Lookup()
+		if _, ok := resultsByID[receipt.ResultID]; !ok {
```

Comment:
Say we have the following chain:
`A <- B <- C(receipt_A) <- D (seal_A,receipt_B) <- E(seal_B)`

The sealing segment for block E is `B <- C <- D <- E` plus `[resultA]`. 

The reason when we add block C, and we learned result A is missing and needed to be queried, is because if B and C don't have result A, the child blocks won't have result A either, that's why we will look up the result from storage, and add it to the missing result list.

But it's important to ensure that we are traversing or adding blocks in the increasing height order. And I would like to add this sanity check here:
When adding a block, this block must be exactly 1 height higher than the highest block. 

Can we add this sanity check?